lib/cadet/jobs/autograder/grading_job.ex

```diff
@@ -0,0 +1,183 @@
+defmodule Cadet.Autograder.GradingJob do
+  @moduledoc """
+  This module contains logic for finding answers to be graded and
+  processing/dispatching them as appropriate
+  """
+  use Cadet, :context
+
+  import Ecto.Query
+
+  alias Cadet.Assessments.{Answer, Assessment, Question, Submission}
+  alias Cadet.Autograder.Utilities
+
+  def grade_all_due_yesterday do
+    for assessment <- Utilities.fetch_assessments_due_yesterday() do
+      assessment.id
+      |> Utilities.fetch_submissions()
+      |> Enum.map(fn %{student_id: student_id, submission: submission} ->
+        if submission do
+          update_submission_status_to_submitted(submission)
+        else
+          insert_empty_submission(%{student_id: student_id, assessment: assessment})
+        end
+      end)
+      |> Enum.each(fn submission ->
+        grade_individual_submission(submission, assessment)
+      end)
+    end
+  end
+
+  @doc """
+  Exposed as public function in case future mix tasks are needed to regrade
+  certain submissions.
+  """
+  def force_grade_individual_submission(
+        submission = %Submission{},
+        assessment = %Assessment{}
+      ) do
+    assessment = preprocess_assessment_for_grading(assessment)
+    grade_individual_submission(submission, assessment, true)
+  end
+
+  # This function requires that assessment questions are already preloaded in sorted
+  # order for autograding to function correctly.
+  defp grade_individual_submission(
+         %Submission{id: submission_id},
+         %Assessment{questions: questions},
+         regrade \\ false
+       ) do
+    answers =
+      Answer
+      |> where(submission_id: ^submission_id)
+      |> order_by(:question_id)
+      |> Repo.all()
+
+    grade_submission_question_answer_lists(
+      submission_id,
+      questions,
+      answers,
+      regrade
+    )
+  end
+
+  defp preprocess_assessment_for_grading(assessment = %Assessment{}) do
+    assessment =
+      if Ecto.assoc_loaded?(assessment.questions) do
+        assessment
+      else
+        Repo.preload(assessment, :questions)
+      end
+
+    Utilities.sort_assessment_questions(assessment)
+  end
+
+  defp insert_empty_submission(%{student_id: student_id, assessment: assessment}) do
+    %Submission{}
+    |> Submission.changeset(%{
+      student_id: student_id,
+      assessment: assessment,
+      status: :submitted
+    })
+    |> Repo.insert!()
+  end
+
+  defp update_submission_status_to_submitted(submission = %Submission{}) do
+    submission
+    |> Submission.changeset(%{status: :submitted})
+    |> Repo.update!()
```

Comment:
None of the updates/inserts in this module should crash since they all don't depend on user input. In the case of the two functions that insert/update submissions, they're only called after user can no longer create submissions (only update status). Inserting answers is also safe since it only happens when no user is able to create submissions. As for updating, this module essentially does fetch -> update. Assumption is that since what I fetch must have been valid, it shouldn't crash either.

In this case, the `update!()` and `insert!()` is more for convenience than anything else.